### PR TITLE
Refactor plugin system

### PR DESCRIPTION
**Pull request by disrupted**
- fix JDBC & S3 sinks not being cleared upon refresh
- more robust plugin loading
	  - detect defining class inside plugin module (using `inspect` instead of `dir`)
	  - handle multiple inheritance
- create abstract base class for all supported plugin types
- create base classes for individual plugin types
- divide Extractor in separate classes to improve usability

nice little added bonus: for the first time Pyright doesn't show any diagnostics whatsoever (including hints)

### CODE REVIEW

File: backend/streams_explorer/core/extractor/extractor.py [hunk 1]
```diff
@@ -9,33 +10,41 @@
...
+    @abstractmethod
     def on_streaming_app_delete(self, config: K8sConfig) -> None:
         ...
 
-    def on_connector_info_parsing(
-        self, info: dict, connector_name: str
-    ) -> KafkaConnector | None:
-        ...
 
+class CronJobExtractor(Extractor):
```
Comment (philipp94831) [inline backend/streams_explorer/core/extractor/extractor.py:47]: In order to name it similarly to StreamsAppExtractor, I would rather call it ProducerAppExtractor

File: backend/streams_explorer/core/extractor/extractor_container.py [hunk 2]
```diff
@@ -37,29 +42,36 @@ def reset(self) -> None:
         for extractor in self.extractors:
             extractor.reset()
 
-    def reset_connector(self) -> None:
+    def reset_connectors(self) -> None:
         for extractor in self.extractors:
-            extractor.reset_connector()
+            if isinstance(extractor, ConnectorExtractor):
```
Comment (philipp94831) [inline backend/streams_explorer/core/extractor/extractor_container.py:47]: For now this is fine but we should move away from using isinstance checks. The best ways to handle this is using proper interfaces where we call methods on or use the visitor pattern

Comment (philipp94831) [inline backend/streams_explorer/core/extractor/extractor_container.py:47]: This is what caused the bug with S3 sink and JDBC, right? I don't see a test that covers this

**Review by philipp94831 — CHANGES_REQUESTED** [review]

File: backend/streams_explorer/plugins.py [hunk 3]
```diff
@@ -46,13 +54,20 @@ def load_plugin(
...
-        ):
-            return plugin_class
+def _find_class(module: ModuleType, base_class: type[T]) -> type[T] | None:
+    members = inspect.getmembers(module, inspect.isclass)
+    if not members:
+        return None
+    if len(members) > 1:  # multiple classes found
+        # exclude classes imported from other modules
+        members = [
+            (name, val) for name, val in members if val.__module__ == module.__name__
```
Comment (VictorKuenstler) [inline backend/streams_explorer/plugins.py:64]: You could use filter here

**Review by VictorKuenstler — CHANGES_REQUESTED** [review]

File: backend/streams_explorer/plugins.py (hunk 3 above)
Comment (disrupted) [reply backend/streams_explorer/plugins.py:64]: that's true, however the use of a simple list comprehension is preferred over `filter` (or `map`) as these are older and often less concise.

I find the list comprehension far easier to read than
```python
members = filter(
    lambda member: member[1].__module__ == module.__name__, members
)
```

File: backend/streams_explorer/core/extractor/extractor_container.py (hunk 2 above)
Comment (disrupted) [reply backend/streams_explorer/core/extractor/extractor_container.py:47]: I agree, thanks for the suggestion about visitor pattern! I will look into it for a further improvement.

Comment (disrupted) [reply backend/streams_explorer/core/extractor/extractor_container.py:47]: I have added a test now

**Review by philipp94831 — APPROVED** [review]

**Pull request merged**